QNN Compilation Support

shubhagr-quic opened the pull request:
1. Infer/Compile API Changes to include --enable_qnn [Optional QNN Config File]
	2. Added qnn_config.json file format.
	3. Added generate_qnn_network_specialization_config.py to create custom_io_config.yaml file for QNN Compilation Step.
	4. Modified utils/constants.py to include QnnConstants required to support QNN compilation.
	5. Updated quick_start.md to include QNN Compilation steps.
	6. Added QNN Compilation utilities in _utils.py

quic-rishinr commented on QEfficient/compile/qnn_compiler.py:
```diff
@@ -0,0 +1,386 @@
+        Returns:
+            :str: Path to compiled ``qpc`` package.
+        """
+        if not (
```
This can be simplified by 
```
if not (self.qnn_config and self.qnn_config.get(QnnConstants.SKIP_QNN_CONVERTOR_STEP_STR)):
```

quic-rishinr commented on QEfficient/compile/qnn_compiler.py:
```diff
@@ -0,0 +1,386 @@
+        """
+        Creates QNN convertor command using provided options.
+
+        ``IMMUTABLE`` Args:
```
Please remove ``IMMUTABLE`` for consistency across codebase. Also include the return statement in doc string

quic-rishinr commented on QEfficient/compile/qnn_compiler.py:
```diff
@@ -0,0 +1,386 @@
+    mxint8: bool = False,
+    full_batch_size=None,
+    qnn_config: Optional[str] = None,
+    **kwargs,
```
kwargs is not utilized anywhere

shubhagr-quic replied:
Added check similar to qaic compiler function:
    if kwargs:
        raise NotImplementedError("Can't handle extra compilation args now!")
Let me know if I should I remove kwargs completely

shubhagr-quic replied:
Added a warning as per Rishin's suggestion.

quic-rishinr commented on QEfficient/compile/qnn_compiler.py:
```diff
@@ -0,0 +1,386 @@
+        compiler_mxfp6_matmul_weights=mxfp6,
+    )
+
+    return qnn_obj.compile()
```
its better to split it as below for better readability 
```
compiled_binary_path = qnn_obj.compile()
return compiled_binary_path 
```

quic-rishinr commented on QEfficient/utils/_utils.py:
```diff
@@ -306,3 +320,59 @@ def get_num_layers_from_config(config):
     return n_layer
+
+
+def execute(process: str, command: str, output_file_path: Optional[str] = None):
```
make it as execute_command for better readability

quic-rishinr commented on QEfficient/utils/generate_qnn_network_specialization_config.py:
```diff
@@ -0,0 +1,141 @@
+    final_dict = {"Input Tensor Configuration": input_nodes_info, "Output Tensor Configuration": output_nodes_info}
+
+    # Save the configuration to a YAML file
+    with open(file_path, "w") as yaml_file:
```
use exception handling

abukhoy reviewed (changes requested):
Please add the test cases for the changed Apis. And try to create a test script for the qnn_complier.py.

shubhagr-quic commented:
> Please add the test cases for the changed Apis. And try to create a test script for the qnn_complier.py.

Modified Unit tests to include QNN compilation path. Added enable_qnn & qnn_config parameter in high_level_testing.json

quic-amitraj commented on QEfficient/cloud/infer.py:
```diff
@@ -60,7 +62,8 @@ def main(
         :hf_token (str): HuggingFace login token to access private repos. ``Defaults to None.``
-
+        :enable_qnn (bool): Enables QNN Compilation. ``Defaults to False.``
+        :qnn_config (str): QNN Config parameters file (if provided).
```
Is this path of the QNN config parameters file, if yes please modify it like-
Path of QNN Config parameters file. ``Defaults is None.``

quic-amitraj commented on QEfficient/compile/compile_helper.py:
```diff
@@ -148,6 +151,8 @@ def compile(
         :mxint8 (bool): Compress Present/Past KV to ``MXINT8`` using ``CustomIO`` config. ``Defaults to False.``
         :custom_io_file_path (str): Path to ``customIO`` file (formatted as a string). ``Defaults to None.``
+        :enable_qnn (bool): Enables QNN Compilation. ``Defaults to False.``
+        :qnn_config (str): QNN Config parameters file (if provided).
```
Same here please.

quic-amitraj commented on QEfficient/compile/qnn_compiler.py:
```diff
@@ -0,0 +1,384 @@
+from QEfficient.utils.logging_utils import logger
+
+
+class QNN:
```
I will be great if you can add docstring for this class.

shubhagr-quic commented on tests/cloud/high_level_testing.json:
```diff
@@ -14,5 +14,7 @@
-    "full_batch_size" : [null,3]
+    "full_batch_size" : [null,3],
+    "enable_qnn" : [false, true],
+    "qnn_config" : [null, "QEfficient/compile/qnn_config.json"]
```
We are using qnn_config to override the default values so "enable_qnn=True and qnn_config=null" is a valid use case.
Also, I have skipped the tests for "enable_qnn=False and qnn_config=QEfficient/compile/qnn_config.json" as it is same to "enable_qnn=False and qnn_config=null"

ochougul commented on QEfficient/compile/qnn_config.json:
better to move this as dictionary in `QEfficient/compile/qnn_compiler.py` python file.
and we parse whenever user passes, is it required to have this saved as json in library?

quic-rishinr replied:
Onkar we need this file as a reference JSON file for users to provide their inputs. Without this file, users won't know the required format for their inputs. That's why we maintain a specific file for configurations.

ochougul commented on QEfficient/utils/generate_qnn_network_specialization_config.py:
move under `QEfficient/compile` dir

shubhagr-quic replied:
Onkar we have kept all the other supporting APIs and constants in the utils folder, so to align with that kept the script to generate network specialization in the same folder. Also the script will undergo further improvements as and when required.

ochougul commented on docs/source/quick_start.md:
````diff
@@ -34,40 +34,55 @@ This is the single e2e CLI API, which takes `model_card` name as input along wit
+python -m QEfficient.cloud.infer --model_name Salesforce/codegen-2B-mono --batch_size 1 --prompt_len 32 --ctx_len 128 --mxfp6 --num_cores 16 --device-group [0,1] --prompt "def fibonacci(n):" --mos 2 --aic_enable_depth_first
+```
+
+For QNN Compilation, add --enable_qnn in the command and an option config file if user wish to override the default parameters.
````
Add that user first needs to setup ENV variable `QNN_SDK_ROOT`

abukhoy reviewed (approved):
Need to update the **gh-pages** after merged.
